rfcs/lower-bounds-calculation.md

```diff
@@ -0,0 +1,217 @@
+# Lower Bounds Calculation
+
+## Summary
+
+We propose adapting lower bounds calculation from Pierce's Local Type Inference paper into the Luau type inference algorithm.
+
+https://www.cis.upenn.edu/~bcpierce/papers/lti-toplas.pdf
+
+## Motivation
+
+There are a number of important scenarios that occur where Luau cannot infer a sensible type without annotations.
+
+Many of these revolve around type variables that occur in contravariant positions.
+
+### Function Return Types
+
+A very common thing to write in Luau is a function to try to find something in some data structure.  These functions habitually return the relevant datum when it is successfully found, or `nil` in the case that it cannot.  For instance:
+
+```lua
+-- A.lua
+function find_first_if(vec, f)
+    for i, e in ipairs(vec) do
+        if f(e) then
+            return i
+        end
+    end
+
+    return nil
+end
+```
+
+This function has two `return` statements: One returns `number` and the other `nil`.  Today, Luau flags this as an error.  We ask authors to add a return annotation to make this error go away.
+
+We would like to automatically infer `find_first_if : <T>({T}, (T) -> boolean) -> number?`.
+
+Higher order functions also present a similar problem.
+
+```lua
+-- B.lua
+function foo(f)
+    f(5)
+    f("string")
+end
+```
+
+There is nothing wrong with the implementation of `foo` here, but Luau fails to typecheck it all the same because `f` is used in an inconsistent way.  This too can be worked around by introducing a type annotation for `f`.
+
+We would like to infer `f : <T...>((number) -> T... & (string) -> T...) -> ()`
+
+## Design
+
+We introduce a new kind of TypeVar, `ConstrainedTypeVar` to represent a TypeVar whose lower bounds are known.  We will never expose syntax for a user to write these types: They only temporarily exist as type inference is being performed.
+
+When unifying some type with a `ConstrainedTypeVar` we _broaden_ the set of constraints that can be placed upon it.
+
+It may help to realize that what we have been doing up until now has been _upper bounds calculation_.
+
+When we `quantify` a function, we will _normalize_ each type and convert each `ConstrainedTypeVar` into a `UnionTypeVar`.
```

Comment:
You mentioned that it's a `UnionTypeVar`, but in 'B.lua' example we want to infer an intersection, how does that resolve?

Reply:
Good catch.  Thanks!